examples/src/bin/ble_bas_peripheral.rs

```diff
@@ -57,27 +60,32 @@ async fn bluetooth_task(sd: &'static Softdevice) {
         info!("advertising done!");
 
         // Run the GATT server on the connection. This returns when the connection gets disconnected.
-        let res = gatt_server::run(&conn, &server, |e| match e {
-            BatteryServiceEvent::BatteryLevelWrite(val) => {
+        let res = gatt_server::run(&conn, |e| match server.on_event(e) {
+            Some(BatteryServiceEvent::BatteryLevelWrite(val)) => {
                 info!("wrote battery level: {}", val);
                 if let Err(e) = server.battery_level_notify(&conn, val + 1) {
                     info!("send notification error: {:?}", e);
                 }
             }
-            BatteryServiceEvent::FooWrite(val) => {
+            Some(BatteryServiceEvent::FooWrite(val)) => {
                 info!("wrote battery level: {}", val);
                 if let Err(e) = server.foo_notify(&conn, val + 1) {
                     info!("send notification error: {:?}", e);
                 }
             }
-            BatteryServiceEvent::BatteryLevelNotificationsEnabled => {
+            Some(BatteryServiceEvent::BatteryLevelNotificationsEnabled) => {
                 info!("battery notifications enabled")
             }
-            BatteryServiceEvent::BatteryLevelNotificationsDisabled => {
+            Some(BatteryServiceEvent::BatteryLevelNotificationsDisabled) => {
                 info!("battery notifications disabled")
             }
-            BatteryServiceEvent::FooNotificationsEnabled => info!("foo notifications enabled"),
-            BatteryServiceEvent::FooNotificationsDisabled => info!("foo notifications disabled"),
+            Some(BatteryServiceEvent::FooNotificationsEnabled) => {
+                info!("foo notifications enabled")
+            }
+            Some(BatteryServiceEvent::FooNotificationsDisabled) => {
+                info!("foo notifications disabled")
+            }
+            None => {}
         })
         .await;
```

Comment:
Not sure how tight a loop this but this is kinda cool because it leaves server available for something like a heapless queue check right here to update server characteristics
```
        if let Some(i) = r.dequeue() {
            unwrap!(server.battery_level_set(i));
        };
```

Reply:
That's an interesting idea. I'm not sure it will work, as the on_event name may be misleading. It is really on_write_event and is only invoked by that. 

I can name it back to on_write, perhaps that is a clearer name after all.